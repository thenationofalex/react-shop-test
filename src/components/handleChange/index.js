export default function handleChange (e) {
  const target = e.target
  const value = target.value
  const name = target.id
  console.log(name, value)
  this.setState({
    [name]: value.toString()
  })
}
