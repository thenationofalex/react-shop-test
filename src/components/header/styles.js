import styled from 'react-emotion'

export const StyledHeader = styled('header')`
  background-color: ${props => props.theme.colors.blue};
  height: 3.35em;
  padding: 0 1.25em;
  position: relative;
  margin-bottom: 1.25em;
  & > h1 {
    margin: 0;
    line-height: 1.5em;
  }
`
