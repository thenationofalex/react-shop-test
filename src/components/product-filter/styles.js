import styled from 'react-emotion'

export const StyledSelect = styled('select')`
  position: absolute;
  top: 1.5em;
  right: 1.25em;
`
