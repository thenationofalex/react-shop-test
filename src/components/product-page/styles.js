import styled from 'react-emotion'

export const StyledPageWrapper = styled('div')`
  max-width: 1320px;
  margin: 0 auto;
`

export const StyledProductWrapper = styled('div')`
  display: flex;
  flex-direction: row;
  flex-flow: wrap;
  & > div {
    width: calc(100% - 2px);
  }
  @media (min-width: 600px) {
    & > div {
      width: calc(50% - 2px);
    }
  }
  @media (min-width: 900px) {
    & > div {
      width: calc(33.33% - 2px);
    }
  }
  @media (min-width: 1200px) {
    & > div {
      width: calc(25% - 2px);
    }
  }
`
