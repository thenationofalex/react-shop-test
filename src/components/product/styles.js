import styled from 'react-emotion'

export const StyledLabel = styled('div')`
  color: #fff;
  font-size: 1em;
  height: 100%;
  max-width: 45%;
  text-align: center;
  text-transform: capitalize;
  line-height: 3.125em;
`

export const StyledProduct = styled('div')`
  padding: 1.25em;
  outline-color #e1e1e1 !important;
  outline: 1px solid;
  margin-top: 1px;
  margin-left: 1px;
  & img {
    width: 100%;
  }
`

export const StyledLabelBlock = styled('div')`
  height: 3.125em;
  display: block;
  vertical-align: middle;
  margin: 0.65em 0;
`

export const StyledSale = styled(StyledLabel)`
  background-color: ${props => props.theme.colors.red};
`

export const StyledExclusive = styled(StyledLabel)`
  background-color: ${props => props.theme.colors.green};
`

export const StyledProductInfo = styled('div')`
  display: flex;
  justify-content: space-between;
  align-items: flex-end;
  & h2 {
    font-size: 1.25rem;
    font-weight: 400;
    margin: 0;
  }
  & p {
    font-size: 2.25rem;
    margin: 0;
  }
`
