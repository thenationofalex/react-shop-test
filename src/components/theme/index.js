export const theme = {
  colors: {
   green: '#009900',
   red: '#ff0000',
   blue: '#def1f4'
  }
}