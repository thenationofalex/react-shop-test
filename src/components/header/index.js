import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import { StyledHeader } from './styles'

class Header extends PureComponent {
  render () {
    return (
      <StyledHeader>
        <h1>{ this.props.title }</h1>
        { this.props.children }
      </StyledHeader>
    )
  }
}

Header.propTypes = {
  title: PropTypes.string,
  children: PropTypes.any
}

export default Header
