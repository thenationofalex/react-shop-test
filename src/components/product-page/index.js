import React, { Fragment, PureComponent } from 'react'
import Head from 'next/head'
import PropTypes from 'prop-types'

import { StyledPageWrapper, StyledProductWrapper } from './styles'
import Header from '../header'
import ProductFilter from '../product-filter'
import Product from '../product'
import handleChange from '../handleChange'

class ProductPage extends PureComponent {
  handleChange = handleChange.bind(this)
  state = { productList: this.props.products }

  componentDidUpdate (prevProps, prevState) {
    if (prevState.size !== this.state.size) {
      let filteredItems = []
      if (this.state.size === 'reset') {
        filteredItems = this.props.products
      } else {
        this.props.products.forEach(product => {
          if (product.size.indexOf(this.state.size) !== -1) {
            filteredItems.push(product)
          }
        })
      }
      this.setState({ productList: filteredItems })
    }
  }
  render () {
    const { title } = this.props
    const { productList } = this.state
    return (
      <StyledPageWrapper>
        <Head>
          <title>{title}</title>
        </Head>
        <Fragment>
          <Header title={title}>
            <ProductFilter options={productList} onChange={this.handleChange} />
          </Header>
          <StyledProductWrapper>
            {
              productList.map(p => <Product key={p.index} {...p} />)
            }
          </StyledProductWrapper>
        </Fragment>
      </StyledPageWrapper>
    )
  }
}

ProductPage.propTypes = {
  title: PropTypes.string.isRequired,
  products: PropTypes.arrayOf(PropTypes.shape({
    index: PropTypes.number,
    isSale: PropTypes.bool,
    isExclusive: PropTypes.bool,
    price: PropTypes.string,
    productImage: PropTypes.string,
    productName: PropTypes.string,
    size: PropTypes.array
  }))
}

export default ProductPage
