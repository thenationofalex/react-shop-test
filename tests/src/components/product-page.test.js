import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import ProductPage from '../../../src/components/product-page'
import { StyledPageWrapper, StyledProductWrapper } from '../../../src/components/product-page/styles'
import { theme } from '../../../src/components/theme'
import products from '../../../static/products.json'

const props = {
  theme
}

describe('Component - Product Page', () => {
  test('Renders', () => {
    const wrapper = shallow(<ProductPage title='Foo' products={products} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('Product Filtering', () => {
    const wrapper = shallow(<ProductPage title='Foo' products={products} />)
    wrapper.instance().setState({ size: 'M' })
    expect(wrapper.state('productList').length).toBe(4)
  })

  test('Product Filtering - Reset', () => {
    const wrapper = shallow(<ProductPage title='Foo' products={products} />)
    wrapper.instance().setState({ size: 'reset' })
    expect(wrapper.state('productList').length).toBe(8)
  })

  test('Product Filtering - Reset', () => {
    const wrapper = shallow(<ProductPage title='Foo' products={products} />)
    wrapper.instance().handleChange({
      target: {
        id: 'size',
        value: 'M'
      }
    })
    expect(wrapper.state('size')).toBe('M')
  })

  test('Styled Component - StyledPageWrapper - Renders', () => {
    const wrapper = shallow(<StyledPageWrapper {...props}>Some content</StyledPageWrapper>)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('Styled Component - StyledProductWrapper - Renders', () => {
    const wrapper = shallow(<StyledProductWrapper {...props}>Some content</StyledProductWrapper>)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
