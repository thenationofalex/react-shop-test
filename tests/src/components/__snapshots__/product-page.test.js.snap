// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Product Page Renders 1`] = `
<Styled(div)>
  <SideEffect(Head)>
    <title>
      Foo
    </title>
  </SideEffect(Head)>
  <Header
    title="Foo"
  >
    <ProductFilter
      onChange={[Function]}
      options={
        Array [
          Object {
            "index": 0,
            "isExclusive": false,
            "isSale": true,
            "price": "$18.88",
            "productImage": "product-1.jpg",
            "productName": "Striped shirt",
            "size": Array [
              "XS",
              "S",
              "L",
              "XL",
            ],
          },
          Object {
            "index": 1,
            "isExclusive": false,
            "isSale": false,
            "price": "$25.44",
            "productImage": "product-2.jpg",
            "productName": "Denim shirt",
            "size": Array [
              "XS",
              "S",
            ],
          },
          Object {
            "index": 2,
            "isExclusive": true,
            "isSale": false,
            "price": "$12.93",
            "productImage": "product-3.jpg",
            "productName": "Plain cotton t-shirt",
            "size": Array [
              "S",
              "M",
            ],
          },
          Object {
            "index": 3,
            "isExclusive": false,
            "isSale": false,
            "price": "$26.32",
            "productImage": "product-4.jpg",
            "productName": "Plain 3/4 sleeve cotton t-shirt",
            "size": Array [
              "XL",
            ],
          },
          Object {
            "index": 4,
            "isExclusive": false,
            "isSale": false,
            "price": "$12.16",
            "productImage": "product-5.jpg",
            "productName": "White dress shirt",
            "size": Array [
              "M",
              "L",
            ],
          },
          Object {
            "index": 5,
            "isExclusive": false,
            "isSale": false,
            "price": "$30.27",
            "productImage": "product-6.jpg",
            "productName": "Long Sleeve Skivvy Top",
            "size": Array [
              "XS",
              "S",
              "M",
            ],
          },
          Object {
            "index": 6,
            "isExclusive": false,
            "isSale": true,
            "price": "$24.26",
            "productImage": "product-7.jpg",
            "productName": "Puffer Vest with Hood",
            "size": Array [
              "M",
              "L",
              "XL",
            ],
          },
          Object {
            "index": 7,
            "isExclusive": true,
            "isSale": false,
            "price": "$17.73",
            "productImage": "product-8.jpg",
            "productName": "Funnel Neck Swing Top",
            "size": Array [
              "XS",
              "S",
              "XL",
            ],
          },
        ]
      }
    />
  </Header>
  <Styled(div)>
    <Product
      index={0}
      isExclusive={false}
      isSale={true}
      key="0"
      price="$18.88"
      productImage="product-1.jpg"
      productName="Striped shirt"
      size={
        Array [
          "XS",
          "S",
          "L",
          "XL",
        ]
      }
    />
    <Product
      index={1}
      isExclusive={false}
      isSale={false}
      key="1"
      price="$25.44"
      productImage="product-2.jpg"
      productName="Denim shirt"
      size={
        Array [
          "XS",
          "S",
        ]
      }
    />
    <Product
      index={2}
      isExclusive={true}
      isSale={false}
      key="2"
      price="$12.93"
      productImage="product-3.jpg"
      productName="Plain cotton t-shirt"
      size={
        Array [
          "S",
          "M",
        ]
      }
    />
    <Product
      index={3}
      isExclusive={false}
      isSale={false}
      key="3"
      price="$26.32"
      productImage="product-4.jpg"
      productName="Plain 3/4 sleeve cotton t-shirt"
      size={
        Array [
          "XL",
        ]
      }
    />
    <Product
      index={4}
      isExclusive={false}
      isSale={false}
      key="4"
      price="$12.16"
      productImage="product-5.jpg"
      productName="White dress shirt"
      size={
        Array [
          "M",
          "L",
        ]
      }
    />
    <Product
      index={5}
      isExclusive={false}
      isSale={false}
      key="5"
      price="$30.27"
      productImage="product-6.jpg"
      productName="Long Sleeve Skivvy Top"
      size={
        Array [
          "XS",
          "S",
          "M",
        ]
      }
    />
    <Product
      index={6}
      isExclusive={false}
      isSale={true}
      key="6"
      price="$24.26"
      productImage="product-7.jpg"
      productName="Puffer Vest with Hood"
      size={
        Array [
          "M",
          "L",
          "XL",
        ]
      }
    />
    <Product
      index={7}
      isExclusive={true}
      isSale={false}
      key="7"
      price="$17.73"
      productImage="product-8.jpg"
      productName="Funnel Neck Swing Top"
      size={
        Array [
          "XS",
          "S",
          "XL",
        ]
      }
    />
  </Styled(div)>
</Styled(div)>
`;

exports[`Component - Product Page Styled Component - StyledPageWrapper - Renders 1`] = `
<div
  className="css-ra2aum"
>
  Some content
</div>
`;

exports[`Component - Product Page Styled Component - StyledProductWrapper - Renders 1`] = `
<div
  className="css-1adyzy8"
>
  Some content
</div>
`;
